source/2015-09-08-meteor-and-three-js.html.markdown

```diff
@@ -0,0 +1,230 @@
+
+~~~
+
+## start render-loop
+
+Template.THREE_scene.onRendered ->
+
+  (...)
+
+  render = =>
+    unless @view.isDestroyed
+      requestAnimationFrame render
+      renderer.render scene, camera
+  render()
+
+  (...)
+
+~~~
+
+## Enter *cursor.observe*
+
+Meteor's *cursor.observe* and *cursor.observeChanges* enables to observe a Mongo-Collection on added, changed or removed documents. This enables us to couple the collection with the *THREE.js*-Scene.
+
+To map document-id's to THREE.js-Objects, we create a simple object that holds all references to THREE.js objects placed to the view (There are other approaches for this like find elements on the scene by name).
+
+The general pattern for this looks like this:
+
+~~~
+
+### start observation
+
+Template.THREE_scene.onRendered ->
+
+  (...)
+
+  # this object will hold references to all THREE.js-objects
+  sceneObjects = {}
+
+  observeHandle = @data.objects?.observeChanges
+    added: (id, fields) ->
+      (...)
+      sceneObjects[id] = threejsObject
+      (...)
+
+    changed: (id, fields) ->
+      # update sceneObjects[id] ....
+      (...)
+
+    removed: (id) ->
+      obj = sceneObjects[id]
+      scene.remove obj
+      delete sceneObjects[id]
+
+  (...)
+
+~~~
+
+### In detail: the *added*-callback
+
+We create simple cubes in this example, but you could also call a constructor dynamically. E.g. you could add a *type*-property to every document and call a matching constructor.
+
+After creation we add the object to the scene and store a reference to this object. We also receive a *fields*-parameter in this callback which holds all properties of the document. We can simply copy these properties onto the THREE-js object with *deepSet*, so that it reflects the state on the database. The function *deepSet* is explained below.
+
+~~~
+
+Template.THREE_scene.onRendered ->
+
+  (...)
+
+  observeHandle = @data.objects?.observeChanges
+    added: (id, fields) ->
+      geometry = new THREE.BoxGeometry 1, 1, 1
+      material = new THREE.MeshLambertMaterial
+      obj = new THREE.Mesh geometry, material
+      scene.add obj
+      # store reference to this object
+      sceneObjects[id] = obj
+      deepSet sceneObjects[id], fields
+
+deepSet = (obj, properties) ->
+  for key, value of properties
+    if _.isObject value
+      deepSet obj[key], value
+    else
+      obj[key] = value
+
+~~~
+
+### In detail: the *change*-callback
+
+The changed-callback will notify us on documents that have been changed in the database. We simply copy the changed properties onto the THREE.js-object (again with *deepSet*).
+
+~~~
+
+    changed: (id, fields) ->
+      # find the referenced object and update all its properties
+      deepSet sceneObjects[id], fields
+
+~~~
+
+### Tidy up
+
+Noticed the *observeHandle*-variable returned from *cursor.observeChanges*? We use this to stop the observation once the template has been destroyed:
```

Comment:
"Noticed" -> "Did you notice how".

Where is that exactly? You could add a hint to where this is stated.